storage/filestore/store.go

```diff
@@ -20,34 +20,34 @@ var (
 
 	mon = monkit.Package()
 
-	_ storage.Blobs = (*Store)(nil)
+	_ storage.Blobs = (*blobStore)(nil)
 )
 
-// Store implements a blob store
-type Store struct {
+// blobStore implements a blob store
+type blobStore struct {
 	dir *Dir
 	log *zap.Logger
 }
 
 // New creates a new disk blob store in the specified directory
-func New(log *zap.Logger, dir *Dir) *Store {
-	return &Store{dir: dir, log: log}
+func New(log *zap.Logger, dir *Dir) *blobStore {
```

Comment:
You may want to return storage.Blobs here to satisfy the linter.

Reply:
Alternatively keep `Store` public as before and just do the other changes.

Reply:
Actually it makes sense to me that storage.Blobs should have `Close` on it directly. I'll do that.

Reply:
I avoid `Close` in interfaces because most of the time, closing is the responsibility of the caller, not the dependency -- or in other words, things using the storage.Blobs, shouldn't close it. And interfaces usually describe requirements for the dependency, rather than a full service itself.

But, I'm not feeling too strongly about it here.

Reply:
I think we might be using storage.Blobs in kind of a weird way here. Is it pretty unusual to have an unexported type which other packages are supposed to use only through an interface? If that's not unusual, what is done in other cases for methods like Close()?

Reply:
I see kind of several ideas going on here:

1. Should other codebase use `*filestore.Store` or `storage.Blobs`.
2. Should `*filestore.Store` be exported.
3. Should `storage.Blobs` have `Close`.

Dissecting my thoughts in too much detail :)

1. The most common and important reason to use an interface in a particular code is to allow variability in the dependency. For example, some tests could use a different implementation - some tests use it to slow down the storagenodes. One reason is to avoid circular dependencies. e.g. A and B are codependent and in different packages and you need to break the cycle, so you end up making one of them an interface (or both). Also it could be to avoid overspecifying the dependency - e.g. making the interface local to a package makes it easier to understand the package as a whole.

For us, the main consideration here is variability - we want to substitute the implementation for tests and caching. Of course, the variability for those could be introduced at some other layer as well.

2. The main reason for unexporting something is to prevent people from depending on things that aren't stable. We do not give any promises about the stability of API-s, except what's in `lib/`. The main reason for exporting is to make it reusable for other things. After specializing the implementation for multiple versions, the reuse definitely isn't a strong argument.

I cannot find a good argument here either way. Our own developers could accidentally use it in places where we shouldn't. But, it also could be used to write some custom filestore implementation for testing.

Of course, when there is a good reason to hide the implementation you need to expose the API somehow, in which case you return an interface; whether it's the same package or a different, probably doesn't matter that much. You can find a bunch of examples from https://golang.org/pkg/io/#Reader.

3. Interfaces often describe what is needed by the dependent. For example, any database dependencies for our services/chores have, don't have a close. There's no reason for them to close it. The closing is all handled where the implementation came from. Of course, there are cases like https://godoc.org/gocloud.dev/blob/driver, where the implementation does need to directly close the Bucket, without knowing how and what implementation looks like. There's always the convenience argument that having `Close` on the actual interface itself rather than augmented in some other way, makes it easier to handle and less verbose.

If the implementation satisfies more things than the main interface then there are two options. e.g. replace it with an anonymous interface `interface { storage.Blobs; Close() error }` alternatively, replace it with a named interface `type Store interface { storage.Blobs; Close() error }`. When you have a lot of different methods with a lot of variation in the methods, then the first one can be nicer - since you don't have to name every single interface. The latter, however, will look slightly nicer in the documentation.

Here, I'm do not have a strong argument either way as well. Convenience is nice, but so is a smaller interface.